hypershift-operator/controllers/machinedeployment/controller.go

```diff
@@ -0,0 +1,279 @@
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/go-logr/logr"
+)
+
+const (
+	// These expose compute information based on the providerSpec input.
+	// This is needed by the autoscaler to foresee upcoming capacity when scaling from zero.
+	// https://github.com/openshift/enhancements/pull/186
+	cpuKey       = "machine.openshift.io/vCPU"
+	memoryKey    = "machine.openshift.io/memoryMb"
+	gpuKey       = "machine.openshift.io/GPU"
+	labelsKey    = "capacity.cluster-autoscaler.kubernetes.io/labels"
+	archLabelKey = "kubernetes.io/arch"
+
+	// Label used to detect if CPO natively supports scale-from-zero
+	controlPlaneOperatorSupportsAWSAutoscalingFromZeroLabel = "io.openshift.hypershift.control-plane-operator-supports-aws-autoscaling-from-zero"
+)
+
+// Reconciler reconciles MachineDeployments for scale-from-zero autoscaling.
+type Reconciler struct {
+	Client             client.Client
+	Log                logr.Logger
+	CredentialsFile    string
+	RegionCache        awsclient.RegionCache
+	InstanceTypesCache InstanceTypesCache
+
+	// For detecting CPO capabilities to skip workaround when CPO supports it natively
+	ImageMetadataProvider   supportutil.ImageMetadataProvider
+	ReleaseProvider         releaseinfo.Provider
+	HypershiftOperatorImage string
+
+	recorder record.EventRecorder
+	scheme   *runtime.Scheme
+}
+
+// SetupWithManager creates a new controller for a manager.
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	_, err := ctrl.NewControllerManagedBy(mgr).
+		For(&clusterv1.MachineDeployment{}).
+		Build(r)
+
+	if err != nil {
+		return fmt.Errorf("failed setting up with a controller manager: %w", err)
+	}
+
+	r.recorder = mgr.GetEventRecorderFor("machinedeployment-controller")
+	r.scheme = mgr.GetScheme()
+	return nil
+}
+
+// Reconcile implements controller runtime Reconciler interface.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.Log.WithValues("machinedeployment", req.Name, "namespace", req.Namespace)
+	logger.V(3).Info("Reconciling")
+
+	machineDeployment := &clusterv1.MachineDeployment{}
+	if err := r.Client.Get(ctx, req.NamespacedName, machineDeployment); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return. Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Ignore deleted MachineDeployments, this can happen when foregroundDeletion
+	// is enabled
+	if !machineDeployment.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
```

Comment:
You should ignore any machinedeployments that are outside of a control plane namespace. I would fetch the namespace of the machinedeployments and immediately return if it doesn't have the `hypershift.openshift.io/hosted-control-plane: "true"` label